### PR TITLE
Add some engine jobs that run with Oracle GraalVM

### LINKED ISSUE #9266: Integrate Oracle GraalVM into Enso CI
Oracle GraalVM (formerly GraalVM EE) is said to offer performance benefits as well as additional functionality (for example **isolates** as needed by #9171). Enso shall not ignore Oracle GraalVM, but continuously work towards using the best of both:
- GraalVM Community
- Oracle GraalVM

Oracle GraalVM comes with [extremely benevolent license](https://www.graalvm.org/faq/) (for a non-open source component) which allows

>  use by any user including commercial and production use. Redistribution is permitted as long as it is not for a fee

As such there is nothing stopping us to use and test against Oracle GraalVM in our CI pipelines. Thus, let's start with **using Oracle GraalVM in our CI**. Having such a CI pipeline set up we can move towards more intricate usages of Oracle GraalVM as outlined by 
- #9171

```[tasklist]
### Tasks
- [x] Set a CI pipeline up that builds and tests engine with Oracle GraalVM
- [x] Execute (engine) benchmarks against Oracle GraalVM
```

### PR DESCRIPTION
**Pull request by Akirathan**
Closes #9266.

### Pull Request Description

Adds `Oracle GraalVM` configuration for some backend jobs. `Oracle GraalVM` jobs run only on Linux so far. The old jobs use `GraalVM CE`.

### Important Notes

- The JDK to download and use is deduced from the `JAVA_VENDOR` environment variable. By default, `GraalVM CE` is used.
- sbt can be started with both GraalVM CE and Oracle GraalVM without any warnings.
  - If you try to start sbt with JDK from a different vendor, but with the same Java version, a warning is printed.

Current list of jobs in the `Engine CI` workflow (these jobs are visible on this PR, because they are scheduled to run on every PR):
- Engine (GraalVM CE) (linux, x86_64)
- Engine (GraalVM CE) (macos, x86_64)
- Engine (GraalVM CE) (windows, x86_64)
- **Engine (Oracle GraalVM) (linux, x86_64)**
- Scala Tests (GraalVM CE) (linux, x86_64)
- Scala Tests (GraalVM CE) (macos, x86_64)
- Scala Tests (GraalVM CE) (windows, x86_64)
- **Scala Tests (Oracle GraalVM) (linux, x86_64)**
- Standard Library Tests (GraalVM CE) (linux, x86_64)
- Standard Library Tests (GraalVM CE) (macos, x86_64)
- Standard Library Tests (GraalVM CE) (windows, x86_64)
- **Standard Library Tests (Oracle GraalVM) (linux x86_64)**
- Verify License Packages (linux, x86_64)

Benchmark Engine workflow (not visible on this PR, cannot schedule manually yet):
- Benchmark Engine (GraalVM CE)
- **Benchmark Engine (Oracle GraalVM)**

Benchmark Standard Libraries workflow (not visible on this PR, cannot schedule manually yet):
- Benchmark Standard Libraries (GraalVM CE)
- **Benchmark Standard Libraries (Oracle GraalVM)**

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] Ensure Oracle GraalVM is downloaded and used in the new jobs.

### CODE REVIEW

**Comment by JaroslavTulach** [comment]
[Oracle GraalVM](https://github.com/enso-org/enso/actions/runs/8192646991/job/22423562824?pr=9322#step:8:233) seems to be downloaded and seems green. The same applies for [scala tests on Oracle GraalVM](https://github.com/enso-org/enso/actions/runs/8192646991/job/22423562500?pr=9322#step:8:253).

But where are _GraalVM CE tests on Linux_? Shouldn't we run both?

File: build/build/src/ci_gen.rs [hunk 1]
```diff
@@ -639,14 +656,31 @@ fn benchmark(name: &str, command_line: &str, timeout_minutes: Option<u32>) -> Re
         wrap_expression(format!("true == inputs.{just_check_input_name}")),
     );
 
-    let mut benchmark_job = RunStepsBuilder::new(command_line)
-        .cleaning(CleaningCondition::Always)
-        .build_job(name, BenchmarkRunner);
-    benchmark_job.timeout_minutes = timeout_minutes;
-    workflow.add_job(benchmark_job);
+    for graal_edition in vec![graalvm::Edition::Community, graalvm::Edition::Enterprise] {
```
Comment (mwu-tow) [inline build/build/src/ci_gen.rs]: Plain array literal can be iterated as well:
```suggestion
    for graal_edition in [graalvm::Edition::Community, graalvm::Edition::Enterprise] {
```

File: build/build/src/ci_gen/job.rs [hunk 2]
```diff
@@ -1,5 +1,7 @@
 use crate::prelude::*;
 
+use heck::ToKebabCase;
```
Comment (mwu-tow) [inline build/build/src/ci_gen/job.rs]: Please run `./run fmt`, so the imports are ordered as per our guidelines.

File: build/build/src/ci_gen/job.rs [hunk 3]
```diff
@@ -157,21 +163,55 @@ impl JobArchetype for VerifyLicensePackages {
 }
 
 #[derive(Clone, Copy, Debug)]
-pub struct ScalaTests;
+pub struct ScalaTests {
+    graal_edition: graalvm::Edition,
```
Comment (mwu-tow) [inline build/build/src/ci_gen/job.rs]: Unless we have a reason to do otherwise (like maintaining some class invariant), we prefer to keep things public.
```suggestion
    pub graal_edition: graalvm::Edition,
```

File: build/build/src/ci_gen/job.rs [hunk 4]
```diff
@@ -157,21 +163,55 @@ impl JobArchetype for VerifyLicensePackages {
...
+}
+
+impl ScalaTests {
+    pub fn with_graal_edition(graal_edition: graalvm::Edition) -> Self {
```
Comment (mwu-tow) [inline build/build/src/ci_gen/job.rs]: By convention, the "constructor" function should be named `new*`. The `with_` functions usually take `self` by value, modify it, and return modified `Self`, allowing chain-style object building.

In this particular case, I'd consider removing the function altogether and have the caller construct the value directly.
i.e. have
```
job::ScalaTests {graal_edition: graalvm::Edition::Community}
```
rather than
```
job::ScalaTests::with_graal_edition(graalvm::Edition::Community)
```

File: build/build/src/ci_gen/job.rs [hunk 5]
```diff
@@ -157,21 +163,55 @@ impl JobArchetype for VerifyLicensePackages {
...
+        Self { graal_edition }
+    }
+}
+
 impl JobArchetype for ScalaTests {
     fn job(&self, target: Target) -> Job {
-        RunStepsBuilder::new("backend test scala")
+        let job_name = format!("Scala Tests ({})", self.graal_edition);
+        let mut job = RunStepsBuilder::new("backend test scala")
             .customize(move |step| vec![step, step::engine_test_reporter(target)])
```
Comment (mwu-tow) [inline build/build/src/ci_gen/job.rs]: I'm not sure but we might want to include the Graal edition as part of the Test Report.
If so, it should be passed to  `engine_test_reporter` along with target.

File: build/build/src/engine.rs [hunk 6]
```diff
@@ -310,9 +311,12 @@ pub async fn deduce_graal(
     build_sbt: &generated::RepoRootBuildSbt,
 ) -> Result<ide_ci::cache::goodie::graalvm::GraalVM> {
     let build_sbt_content = ide_ci::fs::tokio::read_to_string(build_sbt).await?;
+    let graal_edition = env::GRAAL_EDITION.get().map_or(Edition::Community, |e| e);
```
Comment (mwu-tow) [inline build/build/src/engine.rs]: I'd consider deriving `Default` for `graalvm::Edition` and use it here as a fallback.

File: build/ci_utils/src/cache/goodie/graalvm.rs [hunk 7]
```diff
@@ -22,24 +22,77 @@ crate::define_env_var! {
...
+}
+
+impl Into<String> for Edition {
+    fn into(self) -> String {
+        match self {
+            Edition::Community => CE_JAVA_VENDOR.to_string(),
+            Edition::Enterprise => EE_JAVA_VENDOR.to_string(),
+        }
+    }
+}
```
Comment (mwu-tow) [inline build/ci_utils/src/cache/goodie/graalvm.rs:56]: In general, if there's `Display` you get `ToString` for free. Usually then `Into<String>` is not defined, especially if it does the same thing.

File: build/ci_utils/src/cache/goodie/graalvm.rs [hunk 8]
```diff
@@ -22,24 +22,77 @@ crate::define_env_var! {
...
-    Version::find_in_text(line)
+    let edition = if line.contains(CE_JAVA_VENDOR) {
+        Edition::Community
+    } else if line.contains(EE_JAVA_VENDOR) {
+        Edition::Enterprise
+    } else {
+        bail!("Unknown GraalVM edition")
+    };
+    let version = Version::find_in_text(line);
+    version.map(|version| (version, edition)).context("Failed to find GraalVM version")
```
Comment (mwu-tow) [inline build/ci_utils/src/cache/goodie/graalvm.rs]:
```suggestion
    version.map(|version| (version, edition)).context("Failed to find GraalVM version.")
```

File: build/ci_utils/src/cache/goodie/graalvm.rs [hunk 9]
```diff
@@ -22,24 +22,77 @@ crate::define_env_var! {
...
 }
 
-pub async fn find_graal_version() -> Result<Version> {
+async fn find_graal_version() -> Result<(Version, Edition)> {
```
Comment (mwu-tow) [inline build/ci_utils/src/cache/goodie/graalvm.rs]: Why this change? We use `pub` by default.

File: build/ci_utils/src/cache/goodie/graalvm.rs [hunk 10]
```diff
@@ -90,32 +145,53 @@ impl Goodie for GraalVM {
...
+            Edition::Enterprise => {
+                let graal_version_tag = self.graal_version.to_string_core();
+                let url = format!(
+                    "https://download.oracle.com/graalvm/{}/archive/graalvm-jdk-{}_{}_bin.tar.gz",
+                    self.graal_version.major,
+                    graal_version_tag,
+                    self.os_arch_string(),
+                );
+                Box::pin(ready(
+                    Url::parse(&url).context("Failed to parse URL")
```
Comment (mwu-tow) [inline build/ci_utils/src/cache/goodie/graalvm.rs]:
```suggestion
                    Url::parse(&url).context("Failed to parse URL.")
```

File: build/ci_utils/src/cache/goodie/graalvm.rs [hunk 11]
```diff
@@ -90,32 +145,53 @@ impl Goodie for GraalVM {
...
+        let arch_name = match self.arch {
             Arch::X86_64 => "x64",
             Arch::AArch64 => "aarch64",
             other_arch => unimplemented!("Architecture `{}` is not supported!", other_arch),
         };
+        format!("{}-{}", os_name, arch_name)
+    }
+
+    pub fn platform_string(&self) -> String {
+        let Self { graal_version: _graal_version, client: _client , ..} = &self;
```
Comment (mwu-tow) [inline build/ci_utils/src/cache/goodie/graalvm.rs]: I have mixed feelings about using `..` here. The pattern of destructuring `Self { fieds }` is usually used in places like this to force compile error when a new field id added — so the developer is forced to consider whether the new field should be part of the platform string or not.

(this is my thought rather than a change request)

**Review by mwu-tow — COMMENTED** [review]
I am impressed with how well you deal with the build script! 👍 

I have a bunch of comments, mostly stylistic.

File: build/ci_utils/src/cache/goodie/graalvm.rs (hunk 11 above)
Comment (Akirathan) [reply build/ci_utils/src/cache/goodie/graalvm.rs]: Destructuring pattern removed altogether in 10d41f87d5e806591bd18deb80915c5cf6ef5a34

File: build/build/src/ci_gen.rs (hunk 1 above)
Comment (Akirathan) [reply build/build/src/ci_gen.rs]: Fixed in 4d0e5e6ef9f56e56554ce469e37ea6099096f685

File: build/ci_utils/src/cache/goodie/graalvm.rs (hunk 7 above)
Comment (Akirathan) [reply build/ci_utils/src/cache/goodie/graalvm.rs:56]: I have "fixed" this by implementing `From<Edition> for String` trait in 66b03bb850fce28411a134a19d74e4060e1746e9, as suggested by the linter.

File: build/build/src/ci_gen/job.rs (hunk 5 above)
Comment (Akirathan) [reply build/build/src/ci_gen/job.rs]: Good idea. Let's have a different test report for each GraalVM edition - 2ed041938b60b00ccd823a9d2970d67c6698d68c

File: build/build/src/ci_gen/job.rs (hunk 3 above)
Comment (Akirathan) [reply build/build/src/ci_gen/job.rs]: Fixed in da70ac820f48bdcfd9c53b759779003437db22de

File: build/build/src/ci_gen/job.rs (hunk 4 above)
Comment (Akirathan) [reply build/build/src/ci_gen/job.rs]: Fixed in da70ac820f48bdcfd9c53b759779003437db22de

File: build/ci_utils/src/cache/goodie/graalvm.rs (hunk 8 above)
Comment (Akirathan) [reply build/ci_utils/src/cache/goodie/graalvm.rs]: Done in 34923feaa7a32d4c20f93ebc54112a30f73b4ea7

File: build/ci_utils/src/cache/goodie/graalvm.rs (hunk 10 above)
Comment (Akirathan) [reply build/ci_utils/src/cache/goodie/graalvm.rs]: Done in 34923feaa7a32d4c20f93ebc54112a30f73b4ea7

File: build/ci_utils/src/cache/goodie/graalvm.rs (hunk 9 above)
Comment (Akirathan) [reply build/ci_utils/src/cache/goodie/graalvm.rs]: I have JVM background. I like to keep things as hidden as possible. This function was not used outside this module, so I simply removed the `pub` modifier. But if declaring everything public is the correct Rust way, I will revert this change.

Comment (Akirathan) [reply build/ci_utils/src/cache/goodie/graalvm.rs]: Redeclared as public in 859483a83be87d60ce4c4542cfdaaa27e15cd3b5

File: build/build/src/ci_gen/job.rs (hunk 2 above)
Comment (Akirathan) [reply build/build/src/ci_gen/job.rs]: Sorry, my bad, I thought that this is controlled on CI. I have run it manually.

**Review by mwu-tow — APPROVED** [review]

File: .github/workflows/engine-benchmark.yml [hunk 12]
```diff
@@ -13,8 +13,8 @@ on:
         type: boolean
         default: false
 jobs:
-  benchmark-engine:
-    name: Benchmark Engine
+  benchmark-engine-graal-vm-ce:
```
Comment (JaroslavTulach) [inline .github/workflows/engine-benchmark.yml:16]: Won't changing the name affect some visible URLs? E.g. won't the script that downloads the results need a change? If so, then please **don't rename** existing URLs.

File: project/GraalVM.scala [hunk 13]
```diff
@@ -125,10 +130,10 @@ object GraalVM {
       throw new IllegalStateException("GraalVM version check failed")
     }
     val javaVendor = System.getProperty("java.vendor")
-    if (javaVendor != "GraalVM Community") {
+    if (!allowedJavaVendors.contains(javaVendor)) {
```
Comment (JaroslavTulach) [inline project/GraalVM.scala:133]: OK.

**Review by JaroslavTulach — APPROVED** [review]
I am concerned by changes and renames to existing CI actions. I'd prefer to leave them as unchanged as possible. I am afraid the rename will affect other tools that rely on the existing name. Otherwise looks fine.

File: build/build/src/engine.rs (hunk 6 above)
Comment (Akirathan) [reply build/build/src/engine.rs]: Done in 47da8116018919b6585785a9b8fd87aa7a148014

File: .github/workflows/engine-benchmark.yml (hunk 12 above)
Comment (Akirathan) [reply .github/workflows/engine-benchmark.yml:16]: The script that downloads the result relies on *Workflow* (See https://github.com/enso-org/enso/blob/3a3bef0b46e5dd6d2fd6ce4e27139f2519516e84/tools/performance/engine-benchmarks/bench_tool/__init__.py#L25). The workflow corresponds to a `*.yml` file in the `.github/workflows` directory. This PR does not modify workflows, but it modifies only *jobs*. Rename of *jobs* should not have an impact on gathering the results. But thank you for the observation, I will look into these gathering jobs later and ensure that they are gathering the correct results.

Comment (Akirathan) [reply .github/workflows/engine-benchmark.yml:16]: Just to be sure, I have created https://github.com/enso-org/enso/issues/9355 as a follow up issue.
